src/drivers/imu/bmi088/BMI088_accel.cpp

```diff
@@ -0,0 +1,597 @@
+            _reset_wait = hrt_absolute_time() + 10000;
+            _checked_next = 0;
+
+        } else {
+            write_reg(_checked_registers[_checked_next], _checked_values[_checked_next]);
+            // waiting 3ms between register writes seems
+            // to raise the chance of the sensor
+            // recovering considerably
+            _reset_wait = hrt_absolute_time() + 3000;
+        }
+
+        _register_wait = 20;
+    }
+
+    _checked_next = (_checked_next + 1) % BMI088_ACCEL_NUM_CHECKED_REGISTERS;
+}
+
+void
+BMI088_accel::measure()
+{
+    perf_count(_measure_interval);
+
+    if (hrt_absolute_time() < _reset_wait) {
+        // we're waiting for a reset to complete
+        return;
+    }
+
+    struct Report {
+        int16_t     accel_x;
+        int16_t     accel_y;
+        int16_t     accel_z;
+        int16_t     temp;
+    } report;
+
+    /* start measuring */
+    perf_begin(_sample_perf);
+
+    // Checking the status of new data
+    uint8_t status;
+    status = read_reg(BMI088_ACC_STATUS);
+    if (!(status & BMI088_ACC_STATUS_DRDY)) {
+        perf_end(_sample_perf);
+        perf_count(_duplicates);
+        _got_duplicate = true;
+        return;
+    }
+
+    _got_duplicate = false;
+
+    /*
+         * Fetch the full set of measurements from the BMI088 in one pass.
+         */
+    uint8_t index = 0;
+    uint8_t accel_data[8]; // Need an extra byte for the command, and an an extra dummy byte for the read (see section "SPI interface of accelerometer part" of the BMI088 datasheet)
+    accel_data[index] = BMI088_ACC_X_L | DIR_READ;
+
+    const hrt_abstime timestamp_sample = hrt_absolute_time();
+
+    if (OK != transfer(accel_data, accel_data, sizeof(accel_data))) {
+        return;
+    }
+
+    check_registers();
+
+    /* Extracting accel data from the read data */
+    index = 2; // Skip the dummy byte at index=1
+    uint16_t lsb, msb, msblsb;
+
+    lsb = (uint16_t)accel_data[index++];
+    msb = (uint16_t)accel_data[index++];
+    msblsb = (msb << 8) | lsb;
+    report.accel_x = (int16_t)msblsb; /* Data in X axis */
+
+    lsb = (uint16_t)accel_data[index++];
+    msb = (uint16_t)accel_data[index++];
+    msblsb = (msb << 8) | lsb;
+    report.accel_y = (int16_t)msblsb; /* Data in Y axis */
+
+    lsb = (uint16_t)accel_data[index++];
+    msb = (uint16_t)accel_data[index++];
+    msblsb = (msb << 8) | lsb;
+    report.accel_z = (int16_t)msblsb; /* Data in Z axis */
+
+    // Extract the temperature data
+    // Note: the temp sensor data is only updated every 1.28s (see "Register 0x22-0x23 Temperature Sensor Data" section in BMI088 Datasheet)
+    index = 0;
+    accel_data[index] = BMI088_ACC_TEMP_H | DIR_READ;
+    // Need to perform a dummy read, hence the num bytes to read is 3 (plus 1 send byte)
+    if (OK != transfer(accel_data, accel_data, 4)) {
+        return;
+    }
+    index = 2;
+    msb = (uint16_t)accel_data[index++];
+    lsb = (uint16_t)accel_data[index++];
```

Comment:
I'm not sure if I should do a bitwise-and with 0xE0 here since I don't know if it is guaranteed that bits 0-4 will be zero. Although, even if they aren't zero, because we divide by 32, this would be the equivalent of a bitshift to the right, so those bits would be discarded anyway.